src/copy.rs

```diff
@@ -500,40 +513,43 @@ fn copy_past_fork(clipboard: ClipboardType,
                                      });
 
     // Create the data sources and set them as selections.
-    let sources = devices_iter.map(|(device, primary)| {
-                                  let handler = DataSourceHandler::new(data_path.clone(),
-                                                                       should_quit.clone(),
-                                                                       serve_requests.clone());
-                                  let data_source = clipboard_manager.create_data_source(|source| {
-                                                        source.implement(handler, error.clone())
-                                                    })
-                                                    .unwrap();
-
-                                  // If the MIME type is text, offer it in some other common formats.
-                                  if is_text(&mime_type) {
-                                      data_source.offer("text/plain;charset=utf-8".to_string());
-                                      data_source.offer("text/plain".to_string());
-                                      data_source.offer("STRING".to_string());
-                                      data_source.offer("UTF8_STRING".to_string());
-                                      data_source.offer("TEXT".to_string());
-                                  }
-
-                                  data_source.offer(mime_type.clone());
-
-                                  if primary {
-                                      device.set_primary_selection(Some(&data_source));
-                                  } else {
-                                      device.set_selection(Some(&data_source));
-                                  }
-
-                                  // If we need to serve 0 requests, kill the data source right away.
-                                  if let ServeRequests::Only(0) = serve_requests.get() {
-                                      data_source.destroy();
-                                  }
-
-                                  data_source.into()
-                              })
-                              .collect::<Vec<Proxy<_>>>();
+    let mut sources: Vec<Proxy<_>> = Vec::new();
+    for (device, primary) in devices_iter {
+        let handler = DataSourceHandler::new(data_paths.clone(),
+                                            should_quit.clone(),
+                                            serve_requests.clone());
+        let data_source = clipboard_manager.create_data_source(|source| {
+                            source.implement(handler, error.clone())
+                        })
+                        .unwrap();
+
+        for (mime_type, _) in &data_paths {
+            // If the MIME type is text, offer it in some other common formats.
+            if create_multi_text && is_text(&mime_type) {
```

Comment:
I think the way this should work is as follows: with multiple MIME types a text MIME type should still result in offering all these text variations, but it should still be possible to manually offer several data sources for several text MIME types. When a text MIME type is requested, if that specific MIME type was offered, then it should be used. If not, it becomes a little interesting, because we need to pick a fallback out of potentially multiple offered text MIME types. I propose always choosing the first text MIME type passed to `copy_multi()` as the fallback and documenting that.

Reply:
I'll see to it tomorrow

Reply:
Is it enough to remove the `create_multi_text` condition?

Reply:
Or should I collect all given and generated mime types to filter out duplicate before `.offer`ing them?

Reply:
> Is it enough to remove the `create_multi_text` condition?

No: you also need to handle the fallback selection in the `receive` handler.

> Or should I collect all given and generated mime types to filter out duplicate before `.offer`ing them?

Yes, that is a good idea.

Reply:
Commit 285ed1804349777aff227708a50bc0946b4d2567 implements half of this.
However:
- I still have to do the fallback part
- I must address @bugaevc comment about using a RAII guard in make_source, otherwise we get a `Error: Couldn't remove the temporary file` when using it in production

Reply:
>  I propose always choosing the first text MIME type

Isn't it better to leave the current logic, which picks a utf-8 encoding? The only reason I can think of for different content for TEXT mimes is utf-8 vs not utf-8

Reply:
When I'm talking about fallback, I'm talking about the following situation: the crate user does a `copy_multi` with, say, two entries, `MimeType::Specific("text/plain")` and `MimeType::Specific("text/html")`. Then `copy_internal` should recognize a text MIME type and offer the other common text types (`TEXT`, `STRING`, `text/plain;charset=utf-8`, etc.). Then we get a request for, say, `STRING`. Which of the two text sources that the user passed us should be used to satisfy this request? My idea is to use the first text source in the vector. If you've a better idea, please tell me.

Reply:
Right, html is text too. I'll fix it.

Reply:
I implemented the fallback in a slightly different way, because AFAIK there's no way to preserve the order in which they were offered (wayland returns them in random order). In the situation above, we'll get either `text/plain` or `text/html`.
In my opinion, though, the fallback should prefer `text/plain` and similar over `text/html` (like MimeType::Text does)

Reply:
On second thought, isn't it up to the caller to decide what to do, when asking for a specific mime type? If I ask Specific, I probably have a clear idea of what I need. So if it's not available, I might want to react in a different way, for example asking for an other Specific, which might be different from the one we are choosing as fallback here in the library. In my opinion, commits 36687de  and eaf1c02  should be reverted.

Reply:
Thinking more about it, it seems very edge case-prone indeed. For example, what if the caller specifies some MIME type multiple times, or what if the caller specifies `MimeType::Text` along with some specific text MIME type.

I think that perhaps the ideal solution would be for the caller to pass a set of MIME types and some kind of closure which provides the data given a MIME type. But that kind of change is blocked on removing forking (#6).

So I think for now it would be fine to roll back to the simplest solution without fallbacks, as you said. Especially since it works for your use-case.